File: 216. 组合总和 III/Solution.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();
    public List<List<Integer>> combinationSum3(int k, int n) {
        backtracking(n, k, 1, 0);
        return ans;
    }
    private void backtracking(int n, int k, int start, int sum){
        if(sum > n) return;
        if(path.size() == k){
            if(sum == n) ans.add(new ArrayList<>(path));
            return;
        }
        for(int i = start; i <= 9-(k-path.size())+1; i++){
            path.add(i);
            sum += i;
            backtracking(n, k, i+1, sum);
            sum -= i;
            path.removeLast();
        }
    }
}

File: 216. 组合总和 III/Solution.py
class Solution:
    def backtracking(self, start, k, n, path, ans):
        if len(path) == k:
            if sum(path) == n:
                ans.append(path[:])
            return
        for i in range(start, 10):
            path.append(i)
            self.backtracking(i+1, k, n, path, ans)
            path.pop()
        
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ans = []
        path = []
        self.backtracking(1, k, n, path ,ans)
        return ans